File: golang/386_lexical_order/lexical_order.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func lexicalOrder(n int) []int {
	ans := make([]int, n)
	num := 1
	for i := range ans {
		ans[i] = num
		if num*10 <= n {
			num *= 10
		} else {
			for num%10 == 9 || num+1 > n {
				num /= 10
			}
			num++
		}
	}
	return ans
}

func main() {
	tests := []struct {
		n   int
		ans []int
	}{
		{13, []int{1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9}},
		{2, []int{1, 2}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, lexicalOrder(test.n), index)
	}
}

File: cpp/386_lexical_order/lexical_order.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> lexicalOrder(int n) {
    vector<int> ret(n);
    int number = 1;
    for (int i = 0; i < n; i++) {
      ret[i] = number;
      if (number * 10 <= n) {
        number *= 10;
      } else {
        while (number % 10 == 9 || number + 1 > n) {
          number /= 10;
        }
        number++;
      }
    }
    return ret;
  }
};

int main() {
  vector<tuple<int, vector<int>>> tests{
      {13, {1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9}},
      {2, {1, 2}},
  };

  for (auto &[n, ans] : tests) {
    assert(Solution().lexicalOrder(n) == ans);
  }
}